java support with prometheus and dependency management with maven

murali-reddy opened the pull request:
**Issue Ref**: https://github.com/kubeless/kubeless/issues/7
 
**Description**: 

Ths PR adds support for JAVA runtime in Kubeless. PR proposes to use `Maven` as dependency management tool. Since `Maven` is more than dependency management tool, it is also used for packaging and installing build life-cycles.

Since there is not much of code change, I will outline some of the choices and notes:

- Official`openjdk:8-jdk-alpine` image is used as base image, as it is the smallest.
- Maven is not part of the base image, so additional step is used to install Maven.
- below java project layout is used
```
docker/runtime/java
├── Dockerfile
├── Dockerfile.init
├── Makefile
├── function
│   ├── pom.xml
│   └── src
│       └── main
│           └── java
│               └── io
│                   └── kubeless
│                       └── Example.java
├── handler
│   ├── pom.xml
│   └── src
│       └── main
│           └── java
│               └── io
│                   └── kubeless
│                       └── Handler.java
├── params
│   ├── pom.xml
│   └── src
│       └── main
│           └── java
│               └── io
│                   └── kubeless
│                       ├── Context.java
│                       └── Event.java
└── pom.xml
```

- So there are three maven modules. 
    - `params` defines the classes `Event` and `Context` which are used in the function signature. Any Java function  to be used in Kubeless must use the function with signature `public String sayHello(io.kubeless.Event event, io.kubeless.Context context) {` leveraging these classes.
    - `handler` module has the java program that runs webserver and handles the function calls.
    - `function` module is where the user specified java file will be placed.
- java reflection is used to invoke the function
- `mvn package && mvn install` is used for dependency install and compilation by compliation container.

**TODOs**:
 - [x] Ready to review
 - [x] Automated Tests
 - [x] Docs

andresmgot commented on docker/runtime/java/function/src/main/java/io/kubeless/Example.java:
```diff
@@ -0,0 +1,7 @@
+package io.kubeless;
+
+public class Example {
```
probably noob question but can we avoid having a `class` or is it mandatory for some reason?

murali-reddy replied:
Java ia very verbose. Unavoidable.

andresmgot commented on pkg/langruntime/langruntime.go:
```diff
@@ -298,6 +301,11 @@ func (l *Langruntimes) GetCompilationContainer(runtime, funcName string, install
+		command = "cp -r /usr/src/myapp/* /kubeless/ && " +
+			"cp /kubeless/*.java /kubeless/function/src/main/java/io/kubeless/ &&" +
+			"cp /kubeless/function-pom.xml /kubeless/function/pom.xml 2>/dev/null || true &&" +
+			"mvn package && mvn install"
```
should we dump logs to `/dev/termination-log` ?

murali-reddy replied:
Sure, will add, without which there is no insight into why function deployment failed.

murali-reddy commented:
thanks @andresmgot 

@jiparis So with this PR, kubeless can run any Java function even with dependencies. Dependenices are managed through Maven. Maven also used for building the project. So the problem i am having is base image used does not have Maven central repo pom's preinstalled. And every time final container is spun up and `maven package` is called whole pom.xml are downloaded and installed to `.m2`. Which takes around 3-4 min. Is there a way to avoid maven to clone all the pom's.

jiparis commented on docker/runtime/java/handler/src/main/java/io/kubeless/Handler.java:
```diff
@@ -0,0 +1,153 @@
+            Histogram.Timer requestTimer = requestLatency.startTimer();
+            try {
+                requests.inc();
+                Class<?> c = Class.forName("io.kubeless."+className);
```
Since these reflection methods are expensive, I wonder if you could call them and precache the `Method` instance in the `main` method, otherwise they will be called in every function call, affecting performance.

murali-reddy commented:
@jiparis I have addressed all your comments except for suggestion to use `JAX-WS`

I tried to learn a bit on how to use embedd jetty server, but i think i am still not comfartable to make the change as of now. I will make enhancement request. Please share any pointers how to write simple http service with JAZ-WS.

 Having said that image used to run Java function is Kubeless controlled environment so perhaps compability across JRE is not a big concern.

jiparis commented:
Regarding not using com.sun.* packages (sorry I meant JAX-RS, not JAX-WS), there are some good examples out there, like [this](http://nikgrozev.com/2014/10/16/rest-with-embedded-jetty-and-jersey-in-a-single-jar-step-by-step/) or [this](https://www.javacodegeeks.com/2015/03/creating-web-services-and-a-rest-server-with-jax-rs-and-jetty.html). Basically you launch the jetty server in your `main` method and register your JAX-RS service. Using annotations you can easily configure routing and any other aspect.
In any case, as you say, this is a controlled environment, so it may be fine to leave as it is.

murali-reddy commented:
@jiparis Handler part is static, that can be compiled and installed into base image. But the function that user provides is compiled with dependencies at run time. User is free to use any dependencies as long as they are properly specified in the pom.xml that user provides.

> Also, did you consider using gradle instead of maven? My experience is that it's usually faster for both clean and precached builds.
I understand that there is a choice of Maven, gradle, ant etc but we have to choose one. Maven seems fairly popular and familiar to many Java developers. But if there is a need we can revisit later.

>Note that this problem (installing dependencies) is not Java-exclusive, you'll have the same init lag with Gemfiles or Pips.

Yes, but my concern is not on the dependencies that are needed. But maven compile/package operations pull whole central repositories pom.xml into `.m2` which is what i want to avoid if possible.

andresmgot commented on kubeless-non-rbac.jsonnet:
```diff
@@ -160,6 +160,20 @@ local runtime_images ='[
+      {
+        "name": "java1.8",
+        "version": "1.8",
+        "runtimeImage": "muralireddy/java:1.8",
```
btw, don't forget to push this to the `kubeless` repo

murali-reddy commented:
I will land this PR into master. There are couple of improvements as noted in the review

- move to jetty based embedded webserver
- optimize pod startup time by preventing or preinstalling central pom repo xml's

I will open a enhancement request once PR is merged.

murali-reddy commented:
Opened https://github.com/kubeless/kubeless/issues/761 to track enhancements.

@jiparis thanks for your time and review. I will try to replace current implementation with Jetty + JAX-RS and also see if i can preinstall maven central repo pom as part of the enhancement issue.
